### PR TITLE
Pattern matcher improvements

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #97741
* #97931
* __->__ #97740

This adds support for multi-output patterns and example-based
replacements.

Tests/usage are next in this PR stack.

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

**Review by ngimel — APPROVED** [review]

File: torch/_inductor/pattern_matcher.py [hunk 1]
```diff
@@ -404,7 +580,117 @@ def visit(other_node):
...
+    pattern = Converter(gm).run()
+    if not isinstance(pattern, PatternExpr):
+        return MultiOutputPattern(pytree.tree_flatten(pattern)[0])
+    return pattern
+
+
+@torch.no_grad()
+def inference_graph(fn, args):
+    """Build a normalized inference graph, for use with fx_to_pattern"""
+    gm = make_fx(fn, select_decomp_table())(*args)
```
Comment (eellison) [inline torch/_inductor/pattern_matcher.py:641]: Do we need to maintain any invariants here about functionalization ?

**Review by eellison — COMMENTED** [review]
Cool ! Couple questions

File: torch/_inductor/pattern_matcher.py (hunk 1 above)
Comment (eellison) [reply torch/_inductor/pattern_matcher.py:641]: We'll need to do the RNG functionalization at least I think right ?

File: torch/_inductor/pattern_matcher.py [hunk 2]
```diff
@@ -285,107 +387,181 @@ def apply(self, match: Match, graph: torch.fx.Graph, node: torch.fx.Node):
...
+                replacement = [replacement]
+            output_nodes = match.output_nodes()
+            assert len(replacement) == len(output_nodes)
+            for old, new in zip(output_nodes, replacement):
+                if old is None:
+                    assert new is None
+                elif new is None:
+                    old.replace_all_uses_with(None)
+                else:
+                    new.meta.update(old.meta)
```
[the remark this answers is not in the document]
Comment (jansel) [reply torch/_inductor/pattern_matcher.py:424]: I'll add something.

File: torch/_inductor/pattern_matcher.py (hunk 1 above)
Comment (jansel) [reply torch/_inductor/pattern_matcher.py:641]: We will need the behavior here to match AotAutograd for the patterns being used.

File: torch/_inductor/pattern_matcher.py [hunk 3]
```diff
@@ -404,7 +580,117 @@ def visit(other_node):
...
+
+    aot_function(
+        fn,
+        lambda g, i: make_boxed_func(g),
+        partition_fn=record_joint_graph,
+        decompositions=select_decomp_table(),
+    )(*args)
+
+    # remove in/out specs
+    gm.graph._codegen = torch.fx.graph.CodeGen()
```
[the remark this answers is not in the document]
Comment (jansel) [reply torch/_inductor/pattern_matcher.py:666]: These are the same args that compile_fx() uses, so it should get functionalization.

**Pull request closed without merging**